### PR TITLE
Improve sliding-window inference

### PR DESCRIPTION
**Pull request by dongyang0122**
### Description
Improve sliding-window inference with enhanced GPU memory efficiency.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Review by wyli — APPROVED** [review]
pls help double check @Nic-Ma @mingxin-zheng

**Comment by dongyang0122** [comment]
@wyli one of tests failed. I don't know what the root cause is.

**Comment by mingxin-zheng** [comment]
I just re-ran the test `premerge / quick-py3 (macOS-latest) (pull_request)` to make sure it is not related to test resource limitation.

**Comment by mingxin-zheng** [comment]
The code looks fine to me. But I don't have the big picture how it improves the GPU efficiency. @dongyang0122 Can you comment or link to an issue?

File: monai/inferers/utils.py [hunk 1]
```diff
@@ -262,12 +262,17 @@ def sliding_window_inference(
...
+            output_image_list[ss][:, _i : _i + 1, ...] /= _map
+        output_image_list[ss] = output_image_list[ss].to(compute_dtype)
 
     # remove padding if image_size smaller than roi_size
     for ss, output_i in enumerate(output_image_list):
-        if torch.isnan(output_i).any() or torch.isinf(output_i).any():
-            warnings.warn("Sliding window inference results contain NaN or Inf.")
+        for _i in range(output_i.shape[1]):
+            if torch.isnan(output_i[:, _i, ...]).any() or torch.isinf(output_i[:, _i, ...]).any():
+                warnings.warn(f"{_i}th channel of sliding window inference results contains NaN or Inf.")
```
Comment (wyli) [inline monai/inferers/utils.py:275]: shall we remove this check and warning? this is inefficient and the warning message is not very informative in general

**Comment by dongyang0122** [comment]
> The code looks fine to me. But I don't have the big picture how it improves the GPU efficiency. @dongyang0122 Can you comment or link to an issue?

@mingxin-zheng `output_image_list[ss]` is a multi-channel tensor, and `count_map_list.pop(0)` is a single-channel tensor. Direct division between them would create multiple copies of `count_map_list.pop(0)`, which takes big chunks of GPU memory when the channel number is large. Moreover, it seems `torch.isnan(output_i).any()` takes much more GPU memory for 5D-array `output_i`. A for loop would reduce the cost.

File: monai/inferers/utils.py (hunk 1 above)
Comment (mingxin-zheng) [reply monai/inferers/utils.py:275]: I think so too.

Comment (dongyang0122) [reply monai/inferers/utils.py:275]: it is fine to me. but not sure for other applications.

Comment (dongyang0122) [reply monai/inferers/utils.py]: @wyli I have removed this section.

**Pull request merged**